File: explores/graph/1584_min_cost_to_connect_all_points.cpp
#include <climits>
#include <unordered_set>
#include <vector>

class Solution {
   public:
    int minCostConnectPoints(std::vector<std::vector<int>>& points) {
        auto costToConnect = [](const std::vector<int>& p1,
                                const std::vector<int>& p2) -> int {
            return std::abs(p1[0] - p2[0]) + std::abs(p1[1] - p2[1]);
        };

        int minCost = 0;
        std::vector<int> connectedNodes;
        connectedNodes.push_back(0);
        std::unordered_set<int> visited;
        visited.insert(0);
        while (connectedNodes.size() < points.size()) {
            int cost = INT_MAX;
            int nodeToConnect = static_cast<int>(points.size());
            for (int i = 0; i < points.size(); ++i) {
                if (visited.find(i) == visited.end()) {
                    int currentCost =
                        costToConnect(points[connectedNodes[0]], points[i]);
                    for (int j = 1; j < connectedNodes.size(); ++j) {
                        currentCost =
                            std::min(currentCost,
                                     costToConnect(points[connectedNodes[j]],
                                                   points[i]));
                    }
                    if (cost > currentCost) {
                        cost = currentCost;
                        nodeToConnect = i;
                    }
                }
            }
            visited.insert(nodeToConnect);
            connectedNodes.push_back(nodeToConnect);
            minCost += cost;
        }
        return minCost;
    }
};

File: explores/graph/1584_min_cost_to_connect_all_points.py
from typing import List

import UnionFind


class Solution:
    def minCostConnectPoints(self, points: List[List[int]]) -> int:
        def cost_to_connect(x: List[int], y: List[int]) -> int:
            return abs(x[0] - y[0]) + abs(x[1] - y[1])

        edges = []
        for i in range(len(points)):
            for j in range(i + 1, len(points)):
                edges.append([i, j, cost_to_connect(points[i], points[j])])

        total_cost = 0
        n_edges = 0
        uf = UnionFind(len(points))
        edges.sort(key=lambda x: x[-1], reverse=True)
        while n_edges < len(points) - 1:
            t = edges.pop()
            if not uf.connected(t[0], t[1]):
                uf.union(t[0], t[1])
                n_edges += 1
                total_cost += t[2]

        return total_cost
